### PR TITLE
 [FJTN-23] [FJTN-25] edit and delete space

### PR DESCRIPTION
**Pull request by kyrov**
docker-compose.yml
- add port for gRPC service at 50051
gRPC/SpaceServer.go
- add directory for gRPC file
- make editSpaceDetail and deleteSpace service
- add gRPC start server
main.go
- add gRPC start command
model
- add model for spaces and rooms
proto
- add proto file for gRPC service

### CODE REVIEW

File: proto/space.proto [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+//   string faculty = 6;
+//   int64 floor = 7;
+//   string building = 8;
+//   string type = 9;
+//   string head_staff = 10;
+//   repeated string faculty_access_list = 11;
+//   repeated int64 staff_list = 12;
+//   repeated string room_list = 13;
+//   bool is_available = 14;
+// }
```
Comment (phumipatc) [inline proto/space.proto]: remove unused comment section

File: model/space.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	Longitude 			float64 `json:"longitude" gorm:"not null"`
+	Faculty 	 		string `json:"faculty" gorm:"not null"`
+	Floor 		 		int `json:"floor" gorm:"not null"`
+	Building 	 	  	string `json:"building" gorm:"not null"`
+	Type 		 	  	string `json:"type" gorm:"not null"`
+	HeadStaff 	 	  	string `json:"head_staff"`
+	FacultyAccessList 	pq.StringArray `json:"faculty_access_list" gorm:"type:text[]"`
+	StaffList 	 		pq.Int64Array `json:"staff_list" gorm:"type:integer[]"`
+	RoomList 	 		pq.StringArray `json:"room_list" gorm:"type:text[]"`
+	IsAvailable 	 	bool `json:"is_available" gorm:"not null"`
```
Comment (phumipatc) [inline model/space.go]: align the variable type

File: model/room.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import "gorm.io/gorm"
+
+type Room struct {
+	gorm.Model
+	Name 				string `json:"name" gorm:"not null"`
+	Description 		string `json:"description"`
+	RoomNumber 			string `json:"room_number" gorm:"not null"`
+	Capacity 			int `json:"capacity" gorm:"not null"`
+	MinReserveCapacity 	int `json:"min_reserve_capacity" gorm:"not null"`	
+	IsAvailable 		bool `json:"is_available" gorm:"not null"`
```
Comment (phumipatc) [inline model/room.go]: align the variable type

File: main.go [hunk 4]
```diff
@@ -17,6 +19,10 @@ func main() {
 	fmt.Println("Starting server...")
 	
 	model.InitDB()
+	db := model.MainDB
+
+	// start the gRPC server
+	go gRPC.StartGRPCServer(db)
```
Comment (phumipatc) [inline main.go]: Can the parameter be `model.MainDB`? So that no intermediate variable (db) is created.

File: gRPC/SpaceServer.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	space.Latitude = float64(req.Latitude)
+	space.Longitude = float64(req.Longitude)
+	space.Faculty = req.Faculty
+	space.Floor = int(req.Floor)
+	space.Building = req.Building
+	space.Type = req.Type
+	space.HeadStaff = req.HeadStaff
+	space.FacultyAccessList = req.FacultyAccessList
+	space.RoomList = req.RoomList
+	space.IsAvailable = req.IsAvailable
```
Comment (phumipatc) [inline gRPC/SpaceServer.go:24]: If possible, move this section to function `ToSpace`.

```go
func (r *pb.EditSpaceRequest) ToSpace() {
  ...
}
```
or
```go
func ToSpace(r *pb.EditSpaceRequest) {
```

File: gRPC/SpaceServer.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+	lis, err := net.Listen("tcp", ":50051")
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
+	}
+	fmt.Println("gRPC server is running on port 50051")
+	if err := grpcServer.Serve(lis); err != nil {
+		log.Fatalf("failed to serve: %v", err)
+	}
+}
```
Comment (phumipatc) [inline gRPC/SpaceServer.go]: This should be in `grpcServer.go` because others need to use this server too.

File: gRPC/SpaceServer.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+)
+
+type SpaceServer struct {
+	pb.UnimplementedSpaceServiceServer
+	db *gorm.DB
+}
+
+func NewServer(db *gorm.DB) *SpaceServer {
+	return &SpaceServer{db: db}
+}
```
Comment (phumipatc) [inline gRPC/SpaceServer.go]: This should be in `grpcServer.go` because others need to use this server too.

**Comment by phumipatc** [comment]
also please change the PR title to match the template `[FJTN-XX] some description` where xx is task number from JIRA. If there are many tasks involved, match `[FJTN-XX] [FJTN-XX] ...` instead.

**Review by phumipatc — CHANGES_REQUESTED** [review]

File: proto/space.proto (hunk 1 above)
Comment (kyrov) [reply proto/space.proto]: done in new push

**Comment by kyrov** [comment]
Resolve#1
- do all the change request

**Review by phumipatc — APPROVED** [review]
Noice 🥇

**Comment by phumipatc** [comment]
One more thing to do: change PR title
@kyrov

**Pull request merged**